pyrtls.pyi

```diff
@@ -0,0 +1,407 @@
+"""
+# pyrtls: rustls-based modern TLS for Python
+
+[Latest version](https://pypi.org/project/pyrtls) |
+[Documentation](https://pyrtls.readthedocs.io) |
+[CI](https://github.com/djc/pyrtls/actions?query=workflow%3ACI+branch%3Amain)
+
+pyrtls provides bindings to [rustls](https://github.com/rustls/rustls), a modern Rust-based TLS implementation with an API that is
+intended to be easy to use to replace the `ssl` module (but not entirely compatible with it).
+
+In addition to being memory-safe, the library is designed to be more secure by default. As such,
+it does not implement older protocol versions, cipher suites with known security problems, and
+some problematic features of the TLS protocol. For more details, review the [manual](https://docs.rs/rustls/latest/rustls/manual/index.html).
+
+## Warning!
+
+This project is just getting started. While rustls is mature, the Python bindings
+are pretty new and not yet feature-complete.
+"""
+
+from collections.abc import Iterable
+from socket import socket
+
+# lib.rs
+
+class TrustAnchor:
+    def __new__(
+        cls,
+        subject: bytes,
+        subject_public_key_info: bytes,
+        name_constraints: bytes | None,
+    ) -> TrustAnchor: ...
+
+class IoState:
+    def tls_bytes_to_write(self) -> int:
+        """
```

Comment:
What's the value of repeating the docs here?

Reply:
Documentation when you hover over the class/method.

Reply:
Isn't it enough that the actual method has this docstring on it?

Reply:
That doesn't matter because to my knowledge maturin doesn't include docs in the binary, so all of the information needs to be in the stub.

Reply:
I'm still confused, this seems to work:

```python
>>> pyrtls.IoState.tls_bytes_to_write.__doc__
'How many bytes could be written by `Connection.write_tls_into()` if called right now.\nA non-zero value implies that `Connection.wants_write()` would yield `True`.'
```

Reply:
After more testing, it seems like VSCode in particular doesn't use `__doc__`, `__text_signature__`, or `__all__` to provide its completion, only stub files or inline annotations. I can't test on PyCharm or NeoVim since I don't have them, but a lot of Python developers use VSCode so it would be good to support it.

Here's what it looks like without the stub:
![image](https://github.com/user-attachments/assets/bd3c48b1-abe1-4130-a2cf-e1a882d33aca)

And here's it with the stub:
![image](https://github.com/user-attachments/assets/692dff23-ea55-441d-926b-d1f5d5c1e995)

As you can see, it can now detect all the classes and methods in `pyrtls`. The errors are caused by not using the value in the `addr` variable and passing a `str` into `sock.send` when it expects `bytes`.

Reply:
I understand that we need the `pyi` stub to provide type information. My confusion here is specifically about the docstrings -- it seems annoying to duplicate the docs that are in the Rust code (and propagated into the `__doc__` properties) in the stub, and I don't see why what would be needed.

Your screenshot doesn't show any docstrings, so it's not yet obvious to me why duplicating the docstrings is an improvement.

Reply:
That makes sense. Unfortunately, I just tested it again and it seems like VSCode also doesn't pull from the `__doc__` attribute.

When the `pyrtls.pyi` file doesn't have docstrings:
![image](https://github.com/user-attachments/assets/fa32270a-0b80-439c-bfb5-b1f6aa370ee5)

The popup doesn't show a docstring:
![image](https://github.com/user-attachments/assets/c9d972ad-bd8b-4289-8ec4-0d78711e0956)

When the `pyrtls.pyi` file _does_ have docstrings:
![image](https://github.com/user-attachments/assets/9f25fc76-a3b0-4988-be03-1063c3ed9a59)

The popup has the docstring included:
![image](https://github.com/user-attachments/assets/ef2ad6cd-f1b0-4ba5-9b71-8d9dc2b912ec)

Reply:
That seems stupid -- but thanks for demonstrating!